### ISSUE TITLE
dsyev_ and dgemm_ marked for deprecation on MacOS

### ISSUE THREAD

**Issue by remkos** [GenericMappingTools/gmt#8092]
**Description of the problem**

MacOS reports that `dsyev_` from CLAPACK is depracated

**Full script that generated the error**

When doing `make install` on MacOS 14.1.1, I get this warning:

```
gmt/master/src/gmt_vector.c:1098:2: warning: 'dsyev_' is deprecated: first deprecated in macOS 13.3 - The CLAPACK interface is deprecated.  Please compile with -DACCELERATE_NEW_LAPACK to access the new lapack headers. [-Wdeprecated-declarations]
        dsyev_ ( "Vectors", "Upper", &n, a, &lda, w, &wkopt, &lwork, &info );
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/vecLib.framework/Headers/clapack.h:4805:5: note: 'dsyev_' has been explicitly marked deprecated here
int dsyev_(char *__jobz, char *__uplo, __CLPK_integer *__n,
    ^
gmt/master/src/gmt_vector.c:1364:2: warning: 'dgemm_' is deprecated: first deprecated in macOS 13.3 - An updated BLAS interface supporting ILP64 is available.  Please compile with -DACCELERATE_NEW_LAPACK to access the new headers and -DACCELERATE_LAPACK_ILP64 for ILP64 support. [-Wdeprecated-declarations]
        dgemm_ ("t", tr, &na, &nb, &nc, &one, A, &nd, B, &ne, &zero, C, &nf);
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/vecLib.framework/Headers/fortran_blas.h:541:5: note: 'dgemm_' has been explicitly marked deprecated here
int dgemm_(char *transa, char *transb, int *m, int *n, int *k,
    ^
3 warnings generated.
```

Shall we proactively try to mend this? Probably a check in the cmake configuration if we are on a recent darwin?
Not sure though what the new lapack headers are.

**System information**

* Operating system: MacOS 14.1.1
* GMT version (`gmt --version`): 6.5.0_c88738d-dirty_2023.11.09

**Comment by PaulWessel** [comment]
Adding -DACCELERATE_NEW_LAPACK in cmake for testing.  Then we get this

```
/Users/pwessel/UH/RESEARCH/CVSPROJECTS/GMTdev/gmt-dev/src/gmt_vector.c:1333:12: error: conflicting types for 'dgemm_'
extern int dgemm_ (char* tra, char* trb, int* na, int* nb, int* nc, double* alpha, double* a, int *nd, double* b, int *ne, double* beta, double* c, int* nf);
           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/vecLib.framework/Headers/blas_new.h:877:6: note: previous declaration is here
void dgemm_(const char * _Nonnull TRANSA, const char * _Nonnull TRANSB,
     ^
1 error generated.
```

So we need a dummy function that translates between the different kinds of argument. Probably not too hard but also need Cmake language to detect if younger or equal than Monterey(?) on macs?

**Comment by PaulWessel** [comment]
Busy with oncology today but might get to this in the pm unless you beat me to it.

**Comment by remkos** [comment]
This is indicated as marked for deprecation since MacOS 13.3. But it likely means that the newer interface was available since MacOS 13.0.
So your check in Cmake should be either on MacOS >= 13 or Darwin >= 22.

If you want to avoid that we run into problems for MacOS 13.{0,1,2} in case the new interface was then not yet included, then you could also introduce the new interface only for MacOS >= 14 or Darwin >= 23. In the end it is only on MacOS 14 that I noticed this warning.

**Comment by PaulWessel** [comment]
OK, this is the new prototype:

```
void dgemm_(const char * _Nonnull TRANSA, const char * _Nonnull TRANSB,
            const __LAPACK_int * _Nonnull M, const __LAPACK_int * _Nonnull N, const __LAPACK_int * _Nonnull K,
            const double * _Nonnull ALPHA, const double * _Nullable A, const __LAPACK_int * _Nonnull LDA,
            const double * _Nullable B, const __LAPACK_int * _Nonnull LDB,
            const double * _Nonnull BETA, double * _Nullable C, const __LAPACK_int * _Nonnull LDC)
__LAPACK_ALIAS(dgemm)
API_AVAILABLE(macos(13.3), ios(16.4), watchos(9.4), tvos(16.4));
```

So starts with 13.3 as you said.  The old prototype is

`extern int dgemm_ (char* tra, char* trb, int* na, int* nb, int* nc, double* alpha, double* a, int *nd, double* b, int *ne, double* beta, double* c, int* nf);`

Unless I counted wrong they both have 11 arguments so should be one to one.  Will test if a plain cast) from what we have to these crazy things works before having to read docs...

**Comment by remkos** [comment]
And the order appears to be the same as well. So yes, a plain cast should be possible without introducing any routines to reoder arguments.

**Comment by remkos** [comment]
Fixed by #8099
